tests/test_integrated_channels/test_canvas/test_client.py

```diff
@@ -42,17 +43,33 @@ def setUp(self):
             "expires_in": "",
             "access_token": self.access_token
         }
+        self.refresh_token = "refresh_token"
         canvas_factories.CanvasGlobalConfigurationFactory(
-            course_api_path=self.course_api_path,
-            oauth_api_path=self.oauth_api_path,
+            course_api_path=self.course_api_path
         )
         self.enterprise_config = canvas_factories.CanvasEnterpriseCustomerConfigurationFactory(
             client_id=self.client_id,
             secret=self.client_secret,
             canvas_company_id=self.company_id,
             canvas_base_url=self.url_base,
+            refresh_token=self.refresh_token,
         )
 
+    def test_create_client_session_with_oauth_access_key(self):
+        """ Test instantiating the client will fetch and set the session's oauth access key"""
+        with responses.RequestsMock() as rsps:
+            rsps.add(
+                responses.POST,
+                self.oauth_url,
+                json={"access_token": self.access_token},
+                status=200
+            )
+
+            canvas_api_client = CanvasAPIClient(self.enterprise_config)
+            canvas_api_client._create_session()  # pylint: disable=protected-access
+
+            assert canvas_api_client.session.headers['Authorization'] == "Bearer " + self.access_token
```

Comment:
This should just test the `_get_oauth_access_token` and title the method around it as such. You're not really doing anything with the actual client or session here since it's mocked. You're really only testing the functionality of this oauth token function, so just test that directly.

If you want to test the client and session (and you probably should), test what happens under non-ideal conditions. For instance, what happens if the token isn't in the config? Is your code actually handling different status codes correctly? etc. Ideal conditions are useful for verifying your function logic, but it's not very helpful for testing a client functioning as expected. So keep this test to its core functional testing and add some tests here that actually test various responses and what happens if information is missing, because that's extremely valuable in catching bugs before we release this.

Reply:
Fair points, 👍 on the focused testing. I'll add and adjust accordingly.

Reply:
> For instance, what happens if the token isn't in the config?

I think I may have to generalize this test a bit, because as I have it set up, the client doesn't care that the refresh token doesn't exist until it would receive a bad request 400 from Canvas. But also it feels weird because I would be mocking an error response then just testing that the error response was raised. Not sure if that's worth it's own test. Thoughts?

Reply:
Per offline convo:
- I for some reason thought the token was used in local logic, not just directly passing to the canvas API. In which case, testing what happens when it's missing isn't of much use.
- I think there are parts in the code where we're expecting certain errors (there's a part in the client code `except (KeyError, ValueError)`) and I think I'm curious what our fallback is there. We probably do still want to raise on novel errors, because we need to know they exist to handle them better, but not sure it should just cause the whole app to fail, but also silent failures are the worssst.
- I don't think it requires a test necessarily, but just verifying what happens when you are missing parameters. Especially ones we expect to be manually entered by ECS/Customer or dev adds to a config.

Reply:
> verifying what happens when you are missing parameters. Especially ones we expect to be manually entered by ECS/Customer or dev adds to a config.

Manually retested missing client params- it seems like Canvas handles the requests nicely for us. Leaving out either the client id or secret results in `client not found`, no refresh token results in a `refresh token required` error. Grant type is hard-coded so hard to think that it'd be left out, but also get a nicely formatted `grant type not supported` if it's left out.

Technically state is not required but has to match our refresh token request, so there is a possible hole (though I'd argue that if we don't have an enterprise id, we're in a much worse state than not getting the canvas integration). I can add an exception raise if there is no enterprise user associated with the canvas config.

If the canvas base URL isn't there then we won't be able to establish a connection and would result in not a great error, so let me add a sanity check for it before the request. The oauth login path is contained in the app config so I don't think it can go missing?

  > (there's a part in the client code except (KeyError, ValueError)) and I think I'm curious what our fallback is there.

I specified both of those errors because we are already raising for response errors `auth_response.raise_for_status()` which leaves the case of we get a response but it's not in the format that we expect- ie we don't have any json associated with the response or the access token isn't included in the response.